### PR TITLE
schema manager

### PR DESCRIPTION
**Pull request by pounard**
# The plan

- First, this is going to add a schema read API, for introspecting schema.
- Then, and API for altering it.
- Third, normalizing reading and writing to be ISO, and able to diff.
- Maybe a visitor pattern for users, to do whatever they want to do with it, for navigating through the schema as a graph.

# What's working

 - Full MySQL support :heavy_check_mark: 
 - Full MariaDB support :heavy_check_mark: 
 - Full PostgreSQL support :heavy_check_mark: 
 - Partial SQLite support :boar: 

# What's missing

 - SQLServer support (I have a 6.7 linux kernel, and SQL Server crashes with it)
 - SQLite constraints ADD, DROP and ALTER, because they require
 - ~~Allow arbitrary SQL queries during schema transaction, lets say `->query(fn (QueryBuilder): void)`, which does not directly executes the SQL but stacks it into the change log (after all, schema transactions are basically upgrade procedures, and the user probably wants to be able to add data alteration along the way)~~ Done it !
 - ~~ALTER TABLE x MODIFY COLUMN y~~ OK done it!
 - ALTER CONSTRAINT x RENAME TO y
 - Reading indexes
 - Reading ISO to what we are able to write
 - More tests
 - AaAAAAAahhhh COLLATIONS !!! Terrible thing to handle... Some SQL vendors don't want to you to read column collation, some others have terrible names, some other are so dynamic that from one container image to another collation names are not the same...
 - Better column default values handling
 - A method for not sending queries on transaction apply, but a dry-run result instead.
 - Logging !

# Future plans

 - Types normalization, for identifying sequences, serials, identity, etc...
 - Ability for read the schema, then recreate it identically from the definition, required for:
   - Emulation of ADD / DROP / RENAME in SQLite
   - Schema diff

### CODE REVIEW

File: src/Platform/Schema/SQLiteSchemaManager.php [hunk 1]
```diff
@@ -0,0 +1,360 @@
...
+    public function supportsSchema(): bool
+    {
+        return false;
+    }
+
+    #[\Override]
+    public function listDatabases(): array
+    {
+        // @todo Something better?
+        return ['main'];
```
Comment (Lonnytunes) [inline src/Platform/Schema/SQLiteSchemaManager.php:50]: The database filename eventually if you have access to it?

File: docs/content/query/schema.md [hunk 2]
```diff
@@ -0,0 +1,387 @@
...
+$schemaManager = $bridge->getSchemaManager();
+```
+
+`$schemaManager` will be an instance of `MakinaCorpus\QueryBuilder\Schema\SchemaManager`.
+
+## Schema introspection
+
+The schema manager allows reading the current schema:
+
+ - read table data,
```
Comment (SimonMellerin) [inline docs/content/query/schema.md:28]:
```suggestion
 - read table definitions,
```

At the first read, I thought we can retrieve rows from table

File: docs/content/query/schema.md [hunk 3]
```diff
@@ -0,0 +1,387 @@
...
+ - read table columns data,
+ - read table foreign key and reverse foreign key data.
+
+Other implementations will come later.
+
+:::warning
+When possible, this API uses the `information_schema` tables, nevertheless each database vendor
+has its own dialect, even when considering the information schema. Please be aware that each
```
Comment (SimonMellerin) [inline docs/content/query/schema.md:36]:
```suggestion
When possible, this API uses the `information_schema` tables. Nevertheless each database vendor
has its own dialect, even when considering the information schema. Keep in mind that each
```

File: docs/content/query/schema.md [hunk 4]
```diff
@@ -0,0 +1,387 @@
...
+vendor will apply its own access control over information it gives back when reading those
+catalogs, and result may vary depending upon the current user access rights.
```
Comment (SimonMellerin) [inline docs/content/query/schema.md:38]:
```suggestion
catalogs. Result may vary depending upon the current user access rights.
```

File: docs/content/query/schema.md [hunk 5]
```diff
@@ -0,0 +1,387 @@
...
+Default value for `$schema` is always `public`, which is the default name for at
+least PostgreSQL when you create a database from scratch.
+:::
+
+:::warning
+MySQL doesn't support schemas, this parameter will always be ignored when working
+with it, and `listSchemas()` will always return a single value which is `public`.
+:::
+
+Then fetch some table list:
```
Comment (SimonMellerin) [inline docs/content/query/schema.md:69]:
```suggestion
Then fetch table list for a particular database/schema:
```

File: docs/content/query/schema.md [hunk 6]
```diff
@@ -0,0 +1,387 @@
...
+foreach ($schemaManager->listTables('my_database', 'my_schema') as $tableName) {
+    // $tableName is a string
+}
+// $databases is a string array, each value is a database name.
+```
+
+And now, you may also load table information this way:
+
+```php
+$table = $schemaManager->getTable('my_database', 'my_table', 'my_schema');
```
Comment (SimonMellerin) [inline docs/content/query/schema.md:81]: It is a bit confusing to have the 'table' argument before the 'schema'. I understand why you did this, but it is weird to read

Comment (SimonMellerin) [reply docs/content/query/schema.md:81]: May be an intermediate object like "Schema" could solve this ?

```php
$mySchema = $schemaManager->getTable('my_database', 'my_schema');
$table = $mySchema->getTable('my_table');

**Review by SimonMellerin — COMMENTED** [review]
This API seems really nice to use! 

congrats!

File: docs/content/query/schema.md (hunk 6 above)
Comment (pounard) [reply docs/content/query/schema.md:81]: A schema is more or less a namespace, and tables can have relations across different schemas. Most people in real life simply use a unique schema, which is most of the time the default one (`public` in PostgreSQL). Most people won't use the `$schema` argument. That's why it's after.

**Pull request merged**